src/state_circuit/memory.rs

```diff
@@ -129,33 +133,53 @@ impl<F: FieldExt, const NUM_STEPS: usize> Config<F, NUM_STEPS> {
         // Constraints for each step. These are activated when q_memory is nonzero.
         meta.create_gate("Memory operation", |meta| {
             let q_memory = meta.query_selector(q_memory);
+            let q_memory_init = meta.query_selector(q_memory_init);
+            let q_same_value = meta.query_selector(q_same_value);
```

Comment:
Similarly, define a synthetic selector for `q_same_value`. (Separately, I recommend renaming `q_same_value` -> `q_read`). The number of reads varies from proof to proof, and we cannot hard-code the offsets at which we check for a read.

(We will have to move the `one` expression into scope here.)

```suggestion
            let q_same_value = one - flag.clone();
```